src/averageAges.js

```diff
@@ -1,61 +1,74 @@
 'use strict';
 
 /**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
  * @param {object[]} people
  * @param {number} century - optional
  *
  * @return {number}
  */
+
+function calculatePersonsAvgAge(persons) {
+  const calcAges = persons.map(person => person.died - person.born);
+  const personAvgAge = (calcAges.reduce((total, age) =>
+    (total + age)) / calcAges.length);
+
+  return personAvgAge;
+}
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  let menOnly = people.filter(man => (man.sex === 'm'));
+
+  if (century) { // those men who died in the defined century
+    menOnly = menOnly.filter(man => (Math.ceil(man.died / 100) === century));
+  }
+
+  return calculatePersonsAvgAge(menOnly);
 }
 
 /**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
  * @param {object[]} people
  * @param {boolean} withChildren - optional
  *
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const mothers = people.map(eachPerson => eachPerson.mother);
+  let womenOnly = people.filter(woman => woman.sex === 'f');
+
+  if (withChildren) { // checks if person is a mother
+    womenOnly = womenOnly.filter(woman => mothers.includes(woman.name));
+  }
+
+  return calculatePersonsAvgAge(womenOnly);
 }
 
 /**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
  * @param {object[]} people
  * @param {boolean} onlyWithSon - optional
  *
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  // those people only whos mother is listed in 'people'
+  let children = people.filter(child =>
+    people.some(person =>
+      child.mother === person.name));
+
+  if (onlyWithSon) {
+    children = children.filter(person => person.sex === 'm');
+  }
+
+  function calcMotherChildDiff(obj) {
+    const mother = people.find(person => person.name === obj.mother);
+
+    return obj.born - mother.born;
+  }
+
+  const AgeDiff = children.map(person =>
+    calcMotherChildDiff(person));
```

Comment:
just a hint - this syntax would work just fine and shortens the code
```suggestion
  const AgeDiff = children.map(calcMotherChildDiff);
```